### DISCUSSION TITLE
General direct API call for posting meetings

### DISCUSSION THREAD

**Discussion by jcookson-infoblox** [code4recovery/12-step-meeting-list#506] [Q&A] [answered]
**Question :** Is there a direct general API call I can POST JSON-ized data to for importing meetings (of course, it will be site specific -- but after the main site url there should be general tsml plugin REST API url data --- eg, https://my-wp.site.com/{tsl-specific-ap}>/{import-url-here})?
This would be outside using the WP UI, and instead sending a postman-like POST call with the body content being the data to be consumed.  If not, why not & when can we expect is to be made available?

**Comment by joshreisner** [comment, marked as answer]
Hi @jcookson-infoblox - nope that's not a current feature, nor is it on our roadmap. Lots of reasons why:

1. no one's ever asked for it
2. most service entities don't have custom applications that need to sync with their website meeting finder
3. you can already pull data into the plugin via the importer
4. when data flows into the app, either by save, JSON import, or CSV, the address is re-geocoded, and that takes time, especially for bulk requests, so an API service would have to wait a long time between requests
5. sounds like a security nightmare for us to develop a custom REST endpoint
6. wordpress already has a [REST API](https://developer.wordpress.org/rest-api/reference/posts/) which you should be able to use. TSML data is all within the existing data footprint (meetings, locations, and groups are posts, regions and districts are taxonomy terms, everything else is postmeta) 
7. sounds like a lot of new code to support

All that said, you'd be more than welcome to develop (or hire someone to develop) a companion plugin that works with TSML to achieve this objective.

**Reply by jcookson-infoblox** [reply]
ok so here's what I have done .... all of it in _**Python**_

Using the example template.csv parsed out the data into recognizable JSON.  Using the spec data, formatting the entries (example: 6:00 PM == 18:00) as needed so that the data would be ingestible & validating the the JSON output is valid JSON.  Additionally, made customization to the csv meeting resolver to be able to have daily meetings ingested (it would generate the needed 7 meetings for its respective time slot) and accept meeting that we not daily - but on multiple days of the week (ex: Tues/Thurs meetings) and render appropriate meetings into JSON output so that all meetings are in same JSON formatted output.

Currently, the script outputs a JSON file with all the meeting data in it.
Next item is uploading that JSON output to a location that can be added to for feeds to pull from.

What am thinking is a piece that takes whatever csv file meeting data you send it and gets the data to a central space, where an already configured feed can just pick up the info because its hitting a file pointer .... which will not change - only the content within will change.  Am wanting to do this more for getting ew entries into the meeting listings.  Right now the tech knowledge footprint involves knowledge about csv files, client-server setup, plus more .. (it can disenfranchise people who may not feel too techie about things).

**Steps:**
1. resolve meetings from csv (template for otherwise) to json format & output file (done, file name is currently set as export_data.json)
2. upload file to specific place (via data in a config file - to encapsulate the permissions info) -- just started working in it
3. from direct API call or via TSML UI, pull the data into the application via the pre-configured feed. -- investigation phase

**Attached zip file for you to look at,**
[python_tsml.zip](https://github.com/code4recovery/12-step-meeting-list/files/7035706/python_tsml.zip)

**Reply by joshreisner** [reply]
What problem are you trying to solve?

**Reply by jcookson-infoblox** [reply]
Well, seems like most individuals I encounter either have trouble with getting the csv/json right or just getting to the point where they can publish the feed for the website I work with to consume ... I figured there had to be some kind of a simple setup piece ...
Maybe am just over thinking things ... wanting something excessively simple ... I guess

**Reply by joshreisner** [reply]
That helps, we now know generally where the problem is, but it's not a clear picture yet. Could we get some of those individuals on here to describe what aspects of importing with JSON or importing with CSV they found confusing?

**Reply by jcookson-infoblox** [reply]
With google deprecating the sheets API, maybe this can be of some help after all ....
